streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java

```diff
@@ -355,45 +355,65 @@ private void update(final LRUCacheEntry entry) {
 
     private static class NamedCacheMetrics {
         private final StreamsMetricsImpl metrics;
-        private final String groupName;
-        private final Map<String, String> metricTags;
-        private final Map<String, String> allMetricTags;
+
         private final Sensor hitRatioSensor;
+        private final String taskName;
+        private final String cacheName;
 
-        private NamedCacheMetrics(final StreamsMetrics metrics, final String name) {
-            final String scope = "record-cache";
-            final String opName = "hitRatio";
-            final String tagKey = scope + "-id";
-            final String tagValue = ThreadCache.underlyingStoreNamefromCacheName(name);
-            this.groupName = "stream-" + scope + "-metrics";
-            this.metrics = (StreamsMetricsImpl) metrics;
-            this.allMetricTags = ((StreamsMetricsImpl) metrics).tagMap(tagKey, "all",
-                "task-id", ThreadCache.taskIDfromCacheName(name));
-            this.metricTags = ((StreamsMetricsImpl) metrics).tagMap(tagKey, tagValue,
-                "task-id", ThreadCache.taskIDfromCacheName(name));
+        private NamedCacheMetrics(final StreamsMetricsImpl metrics, final String cacheName) {
+            taskName = ThreadCache.taskIDfromCacheName(cacheName);
+            this.cacheName = cacheName;
+            this.metrics = metrics;
 
             // add parent
-            final Sensor parent = this.metrics.registry().sensor(opName, Sensor.RecordingLevel.DEBUG);
-            parent.add(this.metrics.registry().metricName(opName + "-avg", groupName,
-                    "The average cache hit ratio.", allMetricTags), new Avg());
-            parent.add(this.metrics.registry().metricName(opName + "-min", groupName,
-                    "The minimum cache hit ratio.", allMetricTags), new Min());
-            parent.add(this.metrics.registry().metricName(opName + "-max", groupName,
-                    "The maximum cache hit ratio.", allMetricTags), new Max());
+            final Map<String, String> allMetricTags = metrics.tagMap(
+                "record-cache-id", "all",
+                "task-id", taskName
+            );
+            final Sensor threadLevelHitRatioSensor = metrics.threadLevelSensor("hitRatio", Sensor.RecordingLevel.DEBUG);
+            threadLevelHitRatioSensor.add(
+                new MetricName("hitRatio-avg", "stream-record-cache-metrics", "The average cache hit ratio.", allMetricTags),
```

Comment:
The only difference between constructing the MetricName outright and using the registry to make it is that the registry's version will augment your tags with tags from the configs. I verified in https://github.com/apache/kafka/pull/4917/files#diff-06f3e8d1a2ac04162e76b98924c14b70R679 that we don't set tags in the configs, so it should be exactly the same to just construct it directly... right?

Reply:
That's right.

Unfortunately, though, we can't use Metrics to fill in any of those tags, because Metrics can only do global tags. As we have only one Metrics instance for the whole application, it can't even fill in the thread-level tags.

Since we maintain the state necessary to fill in all our own tags when we create metrics, there's no reason to use `Metrics#metricName` and it creates an unnecessary dependency of having a Metrics registry in scope. the `MetricName` constructor, on the other hand, creates no extra dependency. Given two equivalent methods of creating the metric name, it seems better to use the one with less baggage.

Reply:
True, will we ever want to have this ability?  But the change seems fine to me.